### ISSUE TITLE
[Proposal] Promise#then with support to a different executor

### ISSUE THREAD

**Issue by lucasallan** [ruby-concurrency/concurrent-ruby#571]
I recently came across a case where there was 2 executors: global executor for IO operations and a FixedThreadPool for CPU bound ops.

In this case, it would execute some cpu intense operation if the IO operation was successfully.

Something like:

```ruby
Concurrent::Promise.execute { <io-operation> }.then { <cpu operation> }
```

But in order to use both executors, it had to create a new promise in the `then` block.

It could be easily solved if `then` could accept an executor as a parameter.
Example:

```ruby
Concurrent::Promise.execute { <io-operation> }.then(executor: cpu_executor) { <cpu operation> }
``` 

Thoughts?

**Comment by jdantonio** [comment]
That would be very consistent with all the other APIs. 👍 Feel free to make a PR, and thank you!

**Comment by pitr-ch** [comment]
Solved in #572.
@lucasallan fyi: new promises implementation (which will replace current promises) already solves this and I hope to get it released in few months.
